### PR TITLE
feat(iam): `Role.withoutPolicyUpdates()`

### LINKED ISSUE #2985: Allow making iam.IRole functionally "immutable"
**Note: for support questions, please first reference our [documentation](https://docs.aws.amazon.com/cdk/api/latest), then use [Stackoverflow](https://stackoverflow.com/questions/ask?tags=aws-cdk)**. This repository's issues are intended for feature requests and bug reports.

* **I'm submitting a ...**
  - [ ] :beetle: bug report
  - [X] :rocket: feature request
  - [ ] :books: construct library gap
  - [ ] :phone: security issue or vulnerability => Please see [policy](https://github.com/awslabs/aws-cdk/security/policy)
  - [ ] :question: support request => Please see note at the top of this template.

* **What is the current behavior?**
*If the current behavior is a :beetle:bug:beetle:: Please provide the steps to reproduce*

When passing an `IRole` around, permissions can be added to the role under certain conditions (namely, the role must either be owned by the current application, or reside in the same account than the stack granting the additional permissions).

* **What is the expected behavior (or behavior of feature suggested)?**

In certain conditions, a user might want to prohibit new statements being added to a Role policy. The proposition is to offer an `IRole` wrapper that simply ignores attempts to add statements (either directly or through policies) to the role.

Additionally, as imported roles are a common use-case for this, it would be interesting to add an optional `{ ignoreNewStatements = false }` parameter to the `Role.from*` methods.

* **What is the motivation / use case for changing the behavior or adding this feature?**

In certain cases, roles are created by a dedicated organization, and CDK App developers are not allowed to directly modify those. In such cases, attempts to add new statements to imported roles will likely cause in a deployment failure due to insufficient permissions.

In other cases, a user might want to ensure that permissions they hand-crafted for a given role are honored without modification.

This feature means that all places where an optional `IRole` is accepted, the granting behavior for both the "default role" and "user-provided role" cases can be identical, where the construct grants all necessary permissions that are derived from other parameters (example: a SageMaker training job always needs to be able to read/decrypt it's inputs, write/encrypt it's outputs, etc...).

### LINKED ISSUE #4465: CodePipeline/CodeBuild: Maximum policy size of 10240 bytes exceeded for role xxx-role
We are deploying a set of CodePipeline & CodeBuild stacks with AWS CDK and hit the following errors:

```
Maximum policy size of 10240 bytes exceeded for role xxx-CodeBuildRole (Service: AmazonIdentityManagement; Status Code: 409; Error Code: LimitExceeded; Request ID: )
Maximum policy size of 10240 bytes exceeded for role xxx-CodePipelineRole (Service: AmazonIdentityManagement; Status Code: 409; Error Code: LimitExceeded; Request ID: )
Maximum policy size of 10240 bytes exceeded for role xxx-CrossAccountDeployerRole (Service: AmazonIdentityManagement; Status Code: 409; Error Code: LimitExceeded; Request ID:)
```

As you can see, we are trying to use our custom build, pipeline and deployment roles for our setup. 

We have setup the IAM roles so that they are limited to certain resources with certain prefixes in certain regions. The IAM roles are already suitable for pipeline, build and deployment usage. The problem seems to be that CDK adds more policies to the roles until the stack update is cancelled because the resource limit is exceeded.

In our use case we are forced to use hand-crafter CloudFormation templates until we can fully move to CDK. This is currently a blocker for us.

### Reproduction Steps

Creating multiple pipelines with following code:

```javascript

    const pipeline = new codepipeline.Pipeline(stack, `Pipeline`, {
        pipelineName: name,
        role: codePipelineRole(stack), // Import role
        artifactBucket: bucket,
    });
    const source = new codepipeline.Artifact('Source');
    pipeline.addStage({
        stageName: 'Source',
        actions: [
            new codepipeline_actions.CodeCommitSourceAction({
                actionName: 'Source',
                repository,
                output: source,
                role: codeCommitRole(stack), // Import role
            }),
        ]
    });
    const buildOutput = new codepipeline.Artifact();
    pipeline.addStage({
        stageName: 'Build',
        actions: [
            new codepipeline_actions.CodeBuildAction({
                actionName: 'Build',
                input: source,
                project,
                outputs: [buildOutput],
                role: pipeline.role,
            }),
        ]
    });

    pipeline.addStage({
        stageName: 'UpdateDev',
        actions: [
// See https://github.com/aws/aws-cdk/issues/4375
            new MyEcsDeployAction({
                actionName: 'DeployImage',
                clusterName: clusterNameDev,
                serviceName: name,
                input: buildOutput,
                role: crossAccountDeployerRole(stack), // Import role
            }),
        ]
    });

```

### Proposed solution

Two possibilities come to my mind:

- Provide a way to use the role as-is, without adding any new policies and assume it has been already set up properly.

- Another way could possibly be to create new policies rather than inline policies that fill up the limited space very quickly.

### Environment

  - **CLI Version      : 1.12.0
  - **Framework Version: 1.12.0
  - **OS               : MacOS
  - **Language         : TypeScript

--- 

This is :bug: Bug Report

### PR DESCRIPTION
**Pull request by rix0rrr**
Add a new method to `Role` called `withoutPolicyUpdates()`, which returns a wrapper object that will drop all policy updates. This can be used in situations where you want to skip the default IAM permission adding behavior of CDK.

Needs only be used with roles that are created in the same CDK application; for imported roles, supplying `mutable=false` when calling `Role.fromRoleArn()` is sufficient.

Fixes #2985.
Fixes #4465.
Closes #4501.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

## Commit Message
feat(iam): `Role.withoutPolicyUpdates()`

Add a new method to `Role` called `withoutPolicyUpdates()`, which returns a wrapper object that will drop all policy updates. This can be used in situations where you want to skip the default IAM permission adding behavior of CDK.

Needs only be used with roles that are created in the same CDK application; for imported roles, supplying `mutable=false` when calling `Role.fromRoleArn()` is sufficient.

Fixes #2985.
Fixes #4465.
Closes #4501.

### CODE REVIEW

File: packages/@aws-cdk/aws-iam/lib/policy.ts [hunk 1]
```diff
@@ -56,6 +56,22 @@ export interface PolicyProps {
...
+   * (for example, if it remains unattached to an IAM identity or if it has no
+   * statements). This is generally desired behavior, since it prevents
+   * creating invalid--and hence undeployable--CloudFormation templates.
+   *
+   * In cases where you know the policy must be created and it is actually
+   * an error if no statements have been added to it, you can se this to `true`.
+   *
+   * @default false
+   */
+  readonly mustCreate?: boolean;
```
Comment (skinny85) [inline packages/@aws-cdk/aws-iam/lib/policy.ts]: I'm not crazy about this name... How about `lazy?: boolean`?

File: packages/@aws-cdk/aws-iam/lib/policy.ts [hunk 2]
```diff
@@ -160,19 +173,60 @@ export class Policy extends Resource implements IPolicy {
...
     const result = new Array<string>();
 
     // validate that the policy document is not empty
     if (this.document.isEmpty) {
-      result.push('Policy is empty. You must add statements to the policy');
+      if (this.mustCreate) {
+        result.push('Policy created with mustCreate=true is empty. You must add statements to the policy');
+      }
+      if (!this.mustCreate && this.referenceTaken) {
+        result.push('Policy name has been read of empty policy. You must add statements to the policy so it can exist.');
```
Comment (skinny85) [inline packages/@aws-cdk/aws-iam/lib/policy.ts]: I think we can word this a little more clearly... how about: `This Policy has been referenced by another resource, and so it must contain at least one statement`.

File: packages/@aws-cdk/aws-iam/lib/policy.ts [hunk 3]
```diff
@@ -160,19 +173,60 @@ export class Policy extends Resource implements IPolicy {
...
 
     // validate that the policy is attached to at least one principal (role, user or group).
-    if (this.groups.length + this.users.length + this.roles.length === 0) {
-      result.push(`Policy must be attached to at least one principal: user, group or role`);
+    if (!this.isAttached) {
+      if (this.mustCreate) {
+        result.push(`Policy created with mustCreate=true must be attached to at least one principal: user, group or role`);
+      }
+      if (!this.mustCreate && this.referenceTaken) {
+        result.push('Policy name has been read of unattached policy. Attach to at least one principal: user, group or role.');
```
Comment (skinny85) [inline packages/@aws-cdk/aws-iam/lib/policy.ts]: Same here: `This Policy has been referenced by another resource, and so it must be attached to at least one principal: user, group or role`.

**Comment by skinny85** [comment]
> I am not sure I fully understand the use case for this feature, and I couldn't dig those up from the referenced issues. #2985 is about immutable imported roles, #4465 is about maximum policy size limitations and I am not sure I fully understand how immutable roles help here.

The maximum policy size is related to the immutable roles because immutability decreases the size of the statements added to the role when it's passed into constructs that add statements to it. An example from #4465:

```ts
           const myRole = new iam.Role(this, 'Role', { ... });

           // use myRole in a bunch of places...

           // then:
           new codepipeline_actions.CodeBuildAction({
                // ...
                role: myRole,
           });
```

Now, it might happen that passing `myRole` to `CodeBuildAction` will add so many statements that it will exceed the limit. In that case, the customer could do:

```ts
           new codepipeline_actions.CodeBuildAction({
                // ...
                role: new iam.ImmutableRole(myRole),
           });
```

And this way, all statement-adding expressions in `CodeBuildAction` would be effectively ignored, and the customer could work around the "maximum policy size exceeded" error they get.

Hope the motivation for this is a little more clear now!

**Comment by eladb** [comment]
README missing is "changes requested" and also I wanted to make sure the motivation for this, and the use case is clear before it's merged. 

To be honest, the motivation @skinny85 provided above doesn't sounds like a "proper" use case. Maybe it's some kind of an escape hatch to allow people to work around bugs or issues we have in the system, but then I'd argue that a whole new type of `Role` is not the right API.

Let's discuss this before we introduce additional public APIs that don't have a clear use case.

I think trying to write a README entry for this will be a good exercise as it will require you to articulate a reasonable situation where someone would want to use this feature.

**Comment by rix0rrr** [comment]
I think we can chunk this PR smaller. There are some useful changes in there which are separate from the `ImmutableRole` class.

**Comment by rix0rrr** [comment]
Let me try this again as https://github.com/aws/aws-cdk/pull/5608. Although that doesn't fix #4465.

**Comment by rix0rrr** [comment]
Added a README explaining why you might want to use this class.

Pre-emptive FAQ: 

## Can we not do anything else if the policy size is growing too big?

We need to apply judicious use of wildcards then, and we decide not to do IAM policy manipulation because it's scary. Therefore, we have to punt to the user in *some* way. This is that way.

We could make it slightly easier for the user to get an ImmutableRole, maybe instead of 

```ts
new codepipeline.Pipeline(this, 'Pipeline', {
  role: new iam.ImmutableRole(role),
});
```

they could do:

```ts
new codepipeline.Pipeline(this, 'Pipeline', {
  role: role.immutableCopy(),
});
```

But that's about the extent of it, and that's not appreciably different imo.

**Comment by eladb** [comment]
Thanks for adding the README and the FAQ, they are very helpful.

I think this API is much better:

```ts
new codepipeline.Pipeline(this, 'Pipeline', {
  role: role.immutableCopy(),
});
```

As it doesn’t introduce a whole new type and is more discoverable. You just return an IRole this way!. It can also be applied both to owned and imported roles, so there won’t be a difference (maybe even deprecate “mutable: false”).

I am still not sure about the terminology of immutable. Perhaps something more around “role.withoutPolicyUpdates()” or something around “disable automatic policy” which is more in line with the README narrative.

**Comment by rix0rrr** [comment]
Mmyeah. I agree that a `role.withoutPolicyUpdates()` looks nicer at the usage site.

I don't buy adding it to the `IRole`:

* For one, it makes the interface unnecessarily heavy to implement. In Java-like languages (which we are roughly emulating using jsii) there is no way to say "classes that implement `IRole` *automatically* also have these operations available". We're left taking them as an argument, either showing a visible wrapper class (`new Immutable(role: IRole)`, or hiding the name of the wrapper class behind a factory function like `public static makeImmutable(role: IRole): IRole`, which is effectively the same thing).
* For two, I believe that for imports our default of `mutable: true` is probably wrong, and should be `mutable: false`. If you're using existing roles, you're probably looking to use them as-is (99% certainly because they're provisioned to you by your Ops team and you don't have permissions to create new roles), otherwise why wouldn't you be creating new roles? So if we were to put this `withoutPolicyUpdates()` on the `IRole` and rely on people calling that for imports, almost all people doing imports would be writing this:

```ts
const role = Role.fromRoleArn(....).withoutPolicyUpdates();
```

Which is a little silly.

My proposal is:

* Hide the `ImmutableRole` type and put it behind a method `withoutPolicyUpdates()` on the `Role` class.
* In a different PR, switch the `mutable` default of `fromRoleArn()` to `false`.

**Comment by eladb** [comment]
@rix0rrr wrote:

> Hide the ImmutableRole type and put it behind a method withoutPolicyUpdates() on the Role class.

Sounds good.

> In a different PR, switch the mutable default of fromRoleArn() to false.

Yeah I agree that this makes more sense to be honest. And then I would deprecate the name `immutable` and rename this to use similar terminology (something like `policyUpdates: true|false`).

**Comment by skinny85** [comment]
> For two, I believe that for imports our default of mutable: true is probably wrong, and should be mutable: false. If you're using existing roles, you're probably looking to use them as-is (99% certainly because they're provisioned to you by your Ops team and you don't have permissions to create new roles), otherwise why wouldn't you be creating new roles? So if we were to put this withoutPolicyUpdates() on the IRole and rely on people calling that for imports, almost all people doing imports would be writing this:

Are you sure? This  (`mutable: false` - it wasn't called that way, but that's what it effectively was) was the behavior we had originally, and it was changed (to be `mutable: true` by default). I'm pretty sure we had a reason to make that change. Also, I'm not sure reverting to the previous behavior now is backwards-compatible...

**Comment by skinny85** [comment]
I personally think `withoutPolicyUpdates()` is just a roundabout way of saying "I want the role to be immutable", so I'd rather be direct than roundabout, but I don't feel super strongly about this.

**Comment by rix0rrr** [comment]
> Are you sure? This (`mutable: false` - it wasn't called that way, but that's what it effectively was) was the behavior we had originally, and it was changed (to be `mutable: true` by default)

This is where we introduced the `mutable` keyword: https://github.com/aws/aws-cdk/pull/3716 and we went from a mutable imported Role to introducing the `mutable` keyword with default `true`.

We added support for adding policies to imported roles here: https://github.com/aws/aws-cdk/pull/2805. This is where I think we went wrong though.

Motivating issues for that:
https://github.com/aws/aws-cdk/issues/2651 - CodeBuild doesn't work with imported role: solved by not rendering unattached policy.
https://github.com/aws/aws-cdk/issues/2381 - Unable to use SQS event Source with imported role: actual root cause was using the wrong IAM action (but that only became visible when using imported roles)
(2 other issues are referenced that don't actually have anything to do with imported roles)

So it seems like the correct choice to me.

You're right the change is not backwards compatible but I think we'll do everyone a favor in the long run if we rip off the band-aid now. The flag is `@experimental` anyway.

**Comment by skinny85** [comment]
> #2651 - CodeBuild doesn't work with imported role: solved by not rendering unattached policy.

I actually don't agree that this was the error in the general case, as #2651 was a way to workaround an actual bug in CodeBuild (missing `DependsOn` between Policy and Project resources). If you switch this over now, everyone who has a working setup with CodeBuild in a VPC with an imported role will most likely be broken (as we will remove the required permissions CodeBuild needs to [connect to the VPC](https://github.com/aws/aws-cdk/blob/01220cf2139ea801a52d939d4e6e3747720dd3b7/packages/%40aws-cdk/aws-codebuild/lib/project.ts#L944-L973), which is validated at deploy time).

File: packages/@aws-cdk/aws-codebuild/test/test.project.ts [hunk 4]
```diff
@@ -301,4 +301,34 @@ export = {
 
     test.done();
   },
+
+  'can use an ImmutableRole for a Project within a VPC'(test: Test) {
```
Comment (eladb) [inline packages/@aws-cdk/aws-codebuild/test/test.project.ts]:
```suggestion
  'can use role.withoutPolicyUpdates() for a Project within a VPC'(test: Test) {
```

File: packages/@aws-cdk/aws-iam/lib/role.ts [hunk 5]
```diff
@@ -199,44 +216,11 @@ export class Role extends Resource implements IRole {
...
-      }
-    }
-
     const scopeAccount = scopeStack.account;
 
     return options.mutable !== false && accountsAreEqualOrOneIsUnresolved(scopeAccount, roleAccount)
-      ? new MutableImport(scope, id)
-      : new ImmutableImport(scope, id);
+      ? new Import(scope, id)
+      : new ImmutableRole(new Import(scope, id));
```
Comment (eladb) [inline packages/@aws-cdk/aws-iam/lib/role.ts]: should this be `new Import().withoutPolicyUpdates()`?

File: packages/@aws-cdk/aws-iam/lib/role.ts [hunk 6]
```diff
@@ -385,6 +369,19 @@ export class Role extends Resource implements IRole {
...
+  /**
+   * Return a copy of this Role object whose Policies will not be updated
+   *
+   * Use the object returned by this method if you want this Role to be used by
+   * a construct without it automatically updating the Role's Policies.
+   *
+   * If you do, you are responsible for adding the correct statements to the
+   * Role's policies yourself.
+   */
+  public withoutPolicyUpdates(): IRole {
```
Comment (eladb) [inline packages/@aws-cdk/aws-iam/lib/role.ts]: Going back to our conversation, why can't this be in `IRole` and works both for `Role` and imported roles?

Comment (rix0rrr) [reply packages/@aws-cdk/aws-iam/lib/role.ts]: https://github.com/aws/aws-cdk/pull/5569#issuecomment-572538268

1. The interface is already starting to get onerous to implement and I don't want to make it worse. There are 2 different parts to it: "what makes an `IRole` a role", and "operations that can be done to any `IRole`, which are the same for every role anyway". This is part of the 2nd, but I'd argue those shouldn't actually be on `IRole` at all.

2. Even if we put it on `IRole`, my gut says the 99% case will be immutable, hence following up the import by calling this method. We want to optimize for the common case, and adding more verbiage seems the opposite. Either we (ALSO?) add a `withPolicyUpdates()` to switch it back on, or we just leave it as import properties for the `fromRoleArn()`. 

So either we make `IRole` *even heavier* or it doesn't need to be there at all.

Comment (eladb) [reply packages/@aws-cdk/aws-iam/lib/role.ts]: Sure. It’s not that critical.

**Comment by charlocharlie** [comment]
It looks like this PR actually broke `{mutable: false}` for imported roles. Upgrading from 1.20.0 to 1.21.1 causes my imported roles to start taking on policies. I can create an issue if needed.

**Comment by skinny85** [comment]
> It looks like this PR actually broke `{mutable: false}` for imported roles. Upgrading from 1.20.0 to 1.21.1 causes my imported roles to start taking on policies. I can create an issue if needed.

Yes please!

**Comment by charlocharlie** [comment]
Created issue: https://github.com/aws/aws-cdk/issues/5943

**Pull request merged**